core/types/block_ext_test.go

```diff
@@ -106,3 +111,410 @@ func assertDifferentPointers[T any](t *testing.T, a *T, b any) {
 	// Note: no need to check `b` is of the same type as `a`, otherwise
 	// the memory address would be different as well.
 }
+
+func TestBlockExtraGetWith(t *testing.T) {
+	t.Parallel()
+
+	b := &Block{}
+
+	extra := extras.Block.Get(b)
+	require.NotNil(t, extra)
+	assert.Equal(t, &BlockBodyExtra{}, extra)
+
+	extra = &BlockBodyExtra{
+		Version: 1,
+	}
+	extras.Block.Set(b, extra)
+
+	extra = extras.Block.Get(b)
+	wantExtra := &BlockBodyExtra{
+		Version: 1,
+	}
+	assert.Equal(t, wantExtra, extra)
+}
+
+func TestBodyExtraGetWith(t *testing.T) {
+	t.Parallel()
+
+	b := &Body{}
+
+	extra := extras.Body.Get(b)
+	require.NotNil(t, extra)
+	assert.Equal(t, &BlockBodyExtra{}, extra)
+
+	extra = &BlockBodyExtra{
+		Version: 1,
+	}
+	extras.Body.Set(b, extra)
+
+	extra = extras.Body.Get(b)
+	wantExtra := &BlockBodyExtra{
+		Version: 1,
+	}
+	assert.Equal(t, wantExtra, extra)
+}
+
+func TestBodyExtraRLP(t *testing.T) {
+	t.Parallel()
+
+	testTx := NewTransaction(0, common.Address{1}, big.NewInt(2), 3, big.NewInt(4), []byte{5})
+
+	uncle := &Header{ParentHash: common.Hash{6}}
+	uncleExtra := &HeaderExtra{ExtDataHash: common.Hash{7}}
+	SetHeaderExtra(uncle, uncleExtra)
+
+	body := &Body{
+		Transactions: []*Transaction{testTx},
+		Uncles:       []*Header{uncle},
+		Withdrawals:  []*ethtypes.Withdrawal{{Index: 7}}, // ignored
+	}
+	extra := &BlockBodyExtra{
+		Version: 1,
+		ExtData: ptrTo([]byte{2}),
+	}
+	extras.Body.Set(body, extra)
+
+	b, err := rlp.EncodeToBytes(body)
+	require.NoError(t, err)
+
+	gotBody := new(Body)
+	err = rlp.DecodeBytes(b, gotBody)
+	require.NoError(t, err)
+
+	// Check transactions in the following because unexported fields
+	// size and time are set by the RLP decoder.
+	require.Len(t, gotBody.Transactions, 1)
+	gotTx := gotBody.Transactions[0]
+	gotTxJSON, err := gotTx.MarshalJSON()
+	require.NoError(t, err)
+	wantTxJSON, err := testTx.MarshalJSON()
+	require.NoError(t, err)
+	assert.Equal(t, string(wantTxJSON), string(gotTxJSON))
+	gotBody.Transactions = nil
```

Comment:
this seems like a bit of a weird way to check that Transactions and Uncles are equal.
Is there a reflect.DeepEqual or can we just check both Transactions and Uncles separately?

Reply:
I agree. I recommend using `go-cmp` and options like `cmp.Comparer` as [done here in `libevm`](https://github.com/ava-labs/libevm/blob/9f410ecc7ebeb68890e272d2c2b7fe7f37a82f9d/core/types/backwards_compat.libevm_test.go#L107). I'd use `Transaction.Hash()` and `Header.Hash()` as the equality checks.

Reply:
tests refactored, this is included.